### ISSUE TITLE
Remove Ugly Macros

### ISSUE THREAD

**Issue by pete4abw** [pete4abw/libtrash#4]
The original design of `libtrash` used several nested `#define` macros to process the various unlink activities in `open-funs.c`. These are not only difficult to read, but also difficult to maintain. After 20 plus years, perhaps using function pointers in a structure might be a better choice. I'm opening up this issue in the hope that others will come in and provide input and solutions. While these macros work, I just wonder if there is a better way.

```
#define RETURN_FUNCTION_ERROR(err)							\
{											\
	FdOrFp retval;									\
	if (function == FOPEN || function == FREOPEN ||					\
			function == FOPEN64 || function == FREOPEN64)			\
	retval.fp = NULL;								\
	else /* if (function == OPEN || function == OPEN64)*/				\
	retval.fd = -1;									\
	errno = err;									\
	return retval; /* the caller will know which member of the union to access */ 	\
}
```
```
/* This is an ugly macro which invokes the real function with the right arguments and
 * evaluates to its return value: (Note: "real function" refers to the
 * function wrapped by the libtrash function which called us.) */

#define REAL_FUNCTION()										\
	(function == FOPEN ? (*real_fopen) (path, mode_str) :					\
	 (function == FOPEN64 ? (*real_fopen64) (path, mode_str) :				\
	  (function == FREOPEN ? (*real_freopen) (path, mode_str, stream) :			\
	   (function == FREOPEN64 ? (*real_freopen64) (path, mode_str, stream) :		\
	    (function == OPEN ?									\
	     ( (flags & O_CREAT || flags & O_TMPFILE) ? (FILE *) (intptr_t) (*real_open) (path, flags, mode) :	\
	       (FILE *) (intptr_t) (*real_open) (path, flags) ) :							\
	       ( (flags & O_CREAT || flags & O_TMPFILE) ? (FILE *) (intptr_t) (*real_open64) (path, flags) ) ) ) ) ) )
```
```
/* This macro makes the function return the value returned by the function it is emulating, and lets the errno
   value be defined by that function: */

#define RETURN_REAL_FUNCTION()						\
{									\
	FdOrFp retval;							\
	if (function == FOPEN || function == FREOPEN ||			\
			function == FOPEN64 || function == FREOPEN64)	\
		retval.fp = REAL_FUNCTION();				\
	else /* if (function == OPEN || function == OPEN64) */		\
		retval.fd = (intptr_t) REAL_FUNCTION();			\
	return retval;							\
}
```

**Comment by pete4abw** [comment]
Please see new branch, [nomacros](https://github.com/pete4abw/libtrash/tree/nomacros). Review, test, and report. @fweimer-rh . Lots of changes here. I'll wait a bit to merge.

**Comment by pete4abw** [comment]
Fixed in commit cc9ae38, tag v3.9
